src/Oscoin/Storage/Ledger.hs

```diff
@@ -0,0 +1,257 @@
+buildNextBlock ledger time txs = runExceptT $ do
+    (currentTip, tipState) <- ExceptT $ getTipWithState ledger
+    let (newBlock, newState, receipts) = buildBlock (ledgerEvaluator ledger) time tipState txs (blockHash currentTip)
+    lift $ do
+        storeHashContent (ledgerStateStore ledger) newState
+        forM_ receipts $ ReceiptStore.storeReceipt (ledgerReceiptStore ledger)
+    pure newBlock
+
+
+-- | Computes the state and receipts associated with the given block
+-- and adds them to the storage.
+--
+-- The given state must be the state associated with the parent block
+-- of the given block.
+materializeBlockFromState
+    :: (Monad m, Crypto.Hashable crypto tx)
+    => Ledger crypto seal tx output state m
+    -> state
+    -> Block crypto tx (Sealed crypto seal)
+    -> m state
+materializeBlockFromState ledger prevState blk = do
+    let (newState, receipts) = evalBlock (ledgerEvaluator ledger) prevState blk
+    storeHashContent (ledgerStateStore ledger) newState
+    forM_ receipts (ReceiptStore.storeReceipt (ledgerReceiptStore ledger))
+    pure newState
+
+
+-- | Computes the state and receipts associated with the given block
+-- and adds them to the storage.
+--
+-- This is similar to 'materializeBlockFromState' but it looks up the
+-- parent state in the cache and materializes it recursivley if not
+-- present.
+materializeBlock
+    :: (Monad m, Crypto.Hashable crypto tx, HasCallStack)
+    => Ledger crypto seal tx output state m
+    -> Block crypto tx (Sealed crypto seal)
+    -> m (Either (LedgerError crypto) state)
+materializeBlock ledger blk = runExceptT $ do
+    maybeBlockState <- lift $ lookupHashContent (ledgerStateStore ledger) (Crypto.toHashed $ blockStateHash $ blockHeader blk)
+    case maybeBlockState of
+        Just blockState -> pure blockState
+        Nothing          -> do
+            maybeParentBlock <- lift $ BlockStore.lookupBlock (ledgerBlockStoreReader ledger) (parentHash blk)
+            parentBlock <- case maybeParentBlock of
+                Just parentBlock -> pure $ parentBlock
+                Nothing          -> throwError $ BlockNotFound $ parentHash blk
+            parentState <- ExceptT $ materializeBlock ledger parentBlock
+            lift $ materializeBlockFromState ledger parentState blk
+
+
+-- | Computes all states and receipts from the blocks in the
+-- blockchain and adds them to the storage.
+--
+-- The genesis block is applied to the given initial state
+materialize
+    :: (Monad m, Crypto.Hashable crypto tx)
+    => Ledger crypto seal tx output state m
+    -> state
+    -> m ()
+materialize ledger initialState = do
+    genesisBlock <- BlockStore.getGenesisBlock (blockStoreReader ledger)
+    allFollowingBlocks <- BlockStore.getBlocksByParentHash (blockStoreReader ledger) (blockHash genesisBlock)
+    _ <- foldlM (materializeBlockFromState ledger) initialState (genesisBlock : Chrono.toOldestFirst (Chrono.reverse allFollowingBlocks))
```

Comment:
How does that play out with performances and space bounds? We obviously cannot run this function on _all_ the blocks all the time. I guess this is not a concern now, but I suppose what we need to do is to write the ledger state on disk for the adopted, immutable part of the chain, and compute in-memory only the part concerning the immutable part (i.e. the 2016 blocks). Should we track this into a separate ticket?

Reply:
This function is only called when initializing a ledger from a block store. But indeed we should persist the state to not have this run on every boot of the node.

I would be more concerned about `materializeBlock` which is not tail recursive and has running time linear to the number of unevaluated transactions (which might be all transactions).

I think as part of our planning we should elicit what the performance requirements of the ledger are and address them when the need arises.